### PR TITLE
Refactor code for atomic operations

### PR DESCRIPTION
**Pull request by testhound**
Used code generation suggested by @stuartarchibald here: 

https://github.com/numba/numba/commit/9e54c7d78a2d35379724e9e8dfb1b225fd887d96

this pull request refactors many of the atomic operations and reduces code duplication

### CODE REVIEW

File: numba/cuda/cudadecl.py [hunk 1]
```diff
@@ -276,71 +276,36 @@ def generic(self, args, kws):
...
+            if ary.ndim == 1:
+                return signature(ary.dtype, ary, types.intp, ary.dtype)
+            elif ary.ndim > 1:
+                return signature(ary.dtype, ary, idx, ary.dtype)
+    return Cuda_atomic
+
+
+all_numba_types = (types.float64, types.float32,
+                   types.int32, types.uint32,
+                   types.int64, types.uint64)
```
Comment (stuartarchibald) [inline numba/cuda/cudadecl.py:301]: This is applying additional type restriction to the types accepted by `atomic.add` and `atomic.sub`, I think this is probably a good thing as e.g.:
```python
from numba import cuda
import numpy as np

@cuda.jit
def foo(result, values):
    tid = cuda.threadIdx.x
    bid = cuda.blockIdx.x
    bdim = cuda.blockDim.x
    i = (bid * bdim) + tid
    cuda.atomic.add(result, 0, values[i])


arr = np.ones(1, dtype=np.int8)
result = np.ones(1, dtype=np.int8)

foo[1, 1](result, arr)
print(result[0])
```
on mainline results in:
```
numba.cuda.cudadrv.error.NvvmError: Failed to compile

(0) Error: unsupported operation

NVVM_ERROR_COMPILATION
```
this also matches up with the lowering available.

CC @gmarkall

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch, refactor looks good, one query, CC'd @gmarkall for input.

File: numba/cuda/cudadecl.py (hunk 1 above)
Comment (gmarkall) [reply numba/cuda/cudadecl.py:301]: I agree, this is a good thing.

**Review by gmarkall — APPROVED** [review]
Looks good to me, and testsuite passes locally for me.

**Review by stuartarchibald — APPROVED** [review]
Thanks for checking @gmarkall, patch approved!

**Pull request merged**